### PR TITLE
[wasm-ep] Implement EventSource counters support; add an EventPipeSessionOptions builder to TS

### PR DESCRIPTION
**Pull request by lambdageek**
1. Implements support for creating event counters in WebAssembly apps.
   **Important change** adds `Thread.IsInternalThreadStartSupported` and `Thread.InternalUnsafeStart()` to `System.Threading.Thread` that can be used by System.Private.CoreLib to start threads on a runtime where `Thread.IsThreadStartSupported` is false.
   This is used to create the event counters polling thread.
   This is in addition to the native runtime being able to create threads using `mono_thread_create_internal`
2. Coop thread suspend: STW calls `mono_threads_platform_stw_defer_initial_suspend` which postpones suspending a thread until the next GC suspend phase.  Doesn't do anything on most platforms.  On WASM, suspend the main browser thread after all the other threads are suspended. This helps prevent deadlocks where the main thread is suspended while another thread is doing a call that is proxied to the main thread and cannot complete.  By suspending the main thread last, we give it a chance to service the other threads' requests.
2. Adds a `MONO.diagnostics.SessionOptionsBuilder` class that can be used to create an `EventPipeSessionOptions` instance and populate the provider string for an event pipe session.
3. Updated the `browser-eventpipe` sample to create an EventSource and some counters.  The sample is now interactive so you have to click "Start Work" in a browser for things to happen.

There's an outstanding issue https://github.com/dotnet/runtime/issues/69568 that will be investigated in a follow-up PR

### CODE REVIEW

**Comment by lambdageek** [comment]
@dotnet/area-system-runtime FYI, I'm making some changes to some `internal` methods on `System.Thread`.

**Comment by lambdageek** [comment]
@lateralusX @pavelsavara Can I get a review for this PR, please

File: src/mono/sample/wasm/browser-eventpipe/main.js [hunk 1]
```diff
@@ -37,66 +35,86 @@ async function loadRuntime() {
...
+        const ret = await doWork(startWork, stopWork, getIterationsDone);
+        eventSession.stop();
+
+        const filename = "dotnet-wasm-" + makeTimestamp() + ".nettrace";
 
-    if (saveUsingBlob) {
         const blob = eventSession.getTraceBlob();
         const uri = URL.createObjectURL(blob);
         downloadData(uri, filename);
         URL.revokeObjectURL(uri);
```
Comment (kg) [inline src/mono/sample/wasm/browser-eventpipe/main.js:92]: I think this is a race, and it's not necessary

File: src/mono/wasm/runtime/diagnostics.ts [hunk 2]
```diff
@@ -82,30 +82,162 @@ class EventPipeFileSession implements EventPipeSession {
     }
 }
 
+const eventLevel = {
```
Comment (kg) [inline src/mono/wasm/runtime/diagnostics.ts:85]: Why not const enum?

File: src/mono/wasm/runtime/dotnet.d.ts [hunk 3]
```diff
@@ -249,7 +250,39 @@ interface EventPipeSession {
     stop(): void;
     getTraceBlob(): Blob;
 }
+declare const eventLevel: {
```
Comment (kg) [inline src/mono/wasm/runtime/dotnet.d.ts:253]: const enum here too

File: src/mono/wasm/runtime/driver.c [hunk 4]
```diff
@@ -470,7 +470,7 @@ mono_wasm_load_runtime (const char *unused, int debug_level)
 	// We should enable this as part of the wasm build later
 #ifndef DISABLE_THREADS
 	monoeg_g_setenv ("MONO_THREADS_SUSPEND", "coop", 0);
-	monoeg_g_setenv ("MONO_SLEEP_ABORT_LIMIT", "250", 0);
+	monoeg_g_setenv ("MONO_SLEEP_ABORT_LIMIT", "50000", 0);
```
Comment (kg) [inline src/mono/wasm/runtime/driver.c:473]: is this intended? It feels like too large of a value for normal use, but 250 is probably too low

**Review by kg — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/System/Threading/Thread.cs [hunk 5]
```diff
@@ -153,20 +153,24 @@ public Thread(ParameterizedThreadStart start, int maxStackSize)
...
         }
 
-#if !TARGET_BROWSER
-        internal static bool IsThreadStartSupported => true;
-#else
-#if FEATURE_WASM_THREADS
+#if !TARGET_BROWSER || FEATURE_WASM_THREADS
         internal static bool IsThreadStartSupported => true;
+        internal static bool IsInternalThreadStartSupported => true;
+#elif FEATURE_WASM_PERFTRACING
```
Comment (lateralusX) [inline src/libraries/System.Private.CoreLib/src/System/Threading/Thread.cs:159]: Does this have any relationship with the wider, FEATURE_PERFTRACING used to guard all perftracing support?

**Review by lateralusX — APPROVED** [review]
Runtime changes LGTM.

File: src/mono/wasm/runtime/diagnostics.ts (hunk 2 above)
Comment (lambdageek) [reply src/mono/wasm/runtime/diagnostics.ts:85]: Const enum gets completely erased by the typescript compiler - but I actually want `MONO.diagnostics.EventLevel` in JS to have a property for each of the values here.  (So something like https://www.typescriptlang.org/docs/handbook/enums.html#objects-vs-enums)

File: src/libraries/System.Private.CoreLib/src/System/Threading/Thread.cs (hunk 5 above)
Comment (lambdageek) [reply src/libraries/System.Private.CoreLib/src/System/Threading/Thread.cs:159]: I think `FEATURE_PERFTRACING` can be true or false on the wasm threaded runtime, ie it's the general perftracing support. Whereas `FEATURE_WASM_PERFTRACING` is the specific configuration where user threading is disabled.

Comment (lateralusX) [reply src/libraries/System.Private.CoreLib/src/System/Threading/Thread.cs:159]: Yes, FEATURE_PERFTRACING will turn of all diagnostics supports, both in BCL and runtime. So can you build EventPipe/Diagnostic support on WASM without user threading enabled and get something to work?

**Pull request merged**